src/main.cpp

```diff
@@ -4650,17 +4659,23 @@ bool ActivateBestChain(bool fSkipdpow, CValidationState &state, CBlock *pblock)
                 nBlockEstimate = Checkpoints::GetTotalBlocksEstimate(chainParams.Checkpoints());
             // Don't relay blocks if pruning -- could cause a peer to try to download, resulting
             // in a stalled download if the block file is pruned before the request.
-            if (nLocalServices & NODE_NETWORK) {
+            if (nLocalServices & NODE_NETWORK) 
+            {
+                int ht = 0;
+                {
+                    LOCK(cs_main);
+                    ht = chainActive.Height();
+                }
                 LOCK(cs_vNodes);
-                BOOST_FOREACH(CNode* pnode, vNodes)
-                if (chainActive.Height() > (pnode->nStartingHeight != -1 ? pnode->nStartingHeight - 2000 : nBlockEstimate))
-                    pnode->PushInventory(CInv(MSG_BLOCK, hashNewTip));
+                for(CNode* pnode : vNodes)
+                    if (ht > (pnode->nStartingHeight != -1 ? pnode->nStartingHeight - 2000 : nBlockEstimate))
+                        pnode->PushInventory(CInv(MSG_BLOCK, hashNewTip));
             }
             // Notify external listeners about the new tip.
             GetMainSignals().UpdatedBlockTip(pindexNewTip);
             uiInterface.NotifyBlockTip(hashNewTip);
         } //else fprintf(stderr,"initial download skips propagation\n");
-    } while(pindexMostWork != chainActive.Tip());
+    } while(pindexMostWork != GetTipWithLock());
```

Comment:
I checked the bitcoin code - they just use pindexNewTip instead of chainActive.Tip(). That way we may avoid creating an extra function GetTipWithLock()

Reply:
Now using `pindexNewTip`